### PR TITLE
Configure CSP

### LINKED ISSUE #7592: Assess CSP and implement with the new DSL
Rails 5.2 introduced a new DSL for defining a Content Security Policy when handling requests. It needs some assessment and _careful_ implementation/rollout.

Details here: https://bigbinary.com/blog/rails-5-2-adds-dsl-for-configuring-content-security-policy-header

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Closes #7592

Implements a [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) configuration based on recommended defaults, then configures whitelists and exemptions.

**Review notes (and general assessment):** 
- I had to refactor some of the embedded pages code here, as it wasn't compatible with how CSP is configured now
- I had to enable `unsafe_inline` for scripts (which is bad) because we have inline scripts scattered everywhere (which is bad). We should remove all inline scripts and then disable `unsafe_inline` for scripts 
- I had to enable `unsafe_eval` for scripts (which is bad) because of our use of "Angular templates". We should remove/replace them as soon as possible and then disable `unsafe_eval` for scripts
- I had to disable CSP where we use [QZ](https://github.com/openfoodfoundation/ofn-qz) for ticket printing. It's pretty wild. I wonder if we can deprecate/remove it..?

#### What should we test?

General release test?

#### Release notes

Updated CSP configuration

Changelog Category: Technical changes

### CODE REVIEW

File: app/views/layouts/darkswarm.html.haml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-%html
+%html{ "ng-csp": "no-unsafe-eval" }
```
Comment (jibees) [inline app/views/layouts/darkswarm.html.haml:1]: 🙈

**Review by jibees — APPROVED** [review]
Looks good to me. 

Just wondering if those domains should be added as well:
 - `cdn.innocraft.cloud`: it downloads  the Matomo Tag Manager javascript file depending on the container id
 - `m.stripe.network`: it downloads a js file, related to stripe
 - `cdn.loom.com`: it downloads a font file

?

**Comment by Matt-Yorkley** [comment]
@jibees the calls to `cdn.innocraft.cloud` and `m.stripe.network` don't seem to cause any issues with the current configuration. I've been testing with Firefox Laboratory for CSP.

Where did you see `cdn.loom.com` as a font source? I can't see it in the codebase or in any network requests on my end. Is that something being injected by a browser extension?

**Comment by jibees** [comment]
> Where did you see `cdn.loom.com` as a font source? I can't see it in the codebase or in any network requests on my end. Is that something being injected by a browser extension?

Yes, it is actually an extension, read too fast. Sorry.

Thanks!

**Comment by mkllnk** [comment]
>     I had to enable unsafe_inline for scripts (which is bad) because we have inline scripts scattered everywhere (which is bad). We should remove all inline scripts and then disable unsafe_inline for scripts

I agree. Can you create an issue for that?

>     I had to enable unsafe_eval for scripts (which is bad) because of our use of "Angular templates". We should remove/replace them as soon as possible and then disable unsafe_eval for scripts

That's a major re-design from Angular to Reactive Rails, right? Interesting to see this from the security perspective, too. Do we need an issue for this?

>     I had to disable CSP where we use QZ for ticket printing. It's pretty wild. I wonder if we can deprecate/remove it..?

Good question. @oeoeaio implemented that, right? I have no idea if it's still in use but I could imagine so.

File: config/initializers/content_security_policy.rb [hunk 2]
```diff
@@ -10,7 +10,8 @@
   policy.img_src :self, :https, :data, "*.s3.amazonaws.com"
   policy.object_src :none
   policy.frame_ancestors :none
-  policy.script_src :self, :https, :unsafe_inline, :unsafe_eval, "*.stripe.com", "openfoodnetwork.innocraft.cloud"
+  policy.script_src :self, :https, :unsafe_inline, :unsafe_eval, "*.stripe.com", "openfoodnetwork.innocraft.cloud",
+                    "maps.googleapis.com", "maps.gstatic.com", "d2wy8f7a9ursnm.cloudfront.net"
```
Comment (mkllnk) [inline config/initializers/content_security_policy.rb:14]: Hm, I guess that d2wy8f7a9ursnm.cloudfront.net could change without warning...

**Review by mkllnk — APPROVED** [review]
Nice! :sunglasses: 

This is the really important foundation for more security.

File: config/initializers/content_security_policy.rb (hunk 2 above)
Comment (Matt-Yorkley) [reply config/initializers/content_security_policy.rb:14]: As far as I can tell it's Bugsnag's JS endpoint and it's very stable. We've never changed it and it's the URL in their docs :man_shrugging:

Comment (mkllnk) [reply config/initializers/content_security_policy.rb:14]: Right. Sounds relatively safe then.

**Comment by Matt-Yorkley** [comment]
> Can you create an issue for that?

Absolutely! It'll be good to have some clear issues that document what's blocking the improvement of the settings so we have visibility of what needs to be done (even if it doesn't get done quickly).

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

I did a smoke test on the app,
Checkout with Stripe, reports, translations.

Found nothing unusual - I'd say this is good for merge and have a release test before deploy again.

Merging.

**Pull request merged**